libyul/optimiser/LoadResolver.cpp

```diff
@@ -76,3 +102,44 @@ void LoadResolver::tryResolve(
 			if (inScope(*value))
 				_e = Identifier{locationOf(_e), *value};
 }
+
+void LoadResolver::tryEvaluateKeccak(
+	Expression& _e,
+	std::vector<Expression> const& _arguments
+)
+{
+	auto valueOfIdentifier = [&](YulString const& _name) -> optional<u256>
+	{
+		if (m_value.count(_name))
+			if (Literal const* literal = get_if<Literal>(m_value.at(_name).value))
+				return valueOfLiteral(*literal);
+		return nullopt;
+	};
+
+	yulAssert(_arguments.size() == 2, "");
+	Identifier const* memoryKey =  std::get_if<Identifier>(&_arguments.at(0));
+	Identifier const* offset = std::get_if<Identifier>(&_arguments.at(1));
+
+	if (!memoryKey || !offset) return;
+
+	if (
+		auto memoryValue = util::valueOrNullptr(m_memory, memoryKey->name);
+		memoryValue &&
+		inScope(*memoryValue)
+	)
+	{
+		optional<u256> memoryContent = valueOfIdentifier(*memoryValue);
+		optional<u256> memoryOffset = valueOfIdentifier(offset->name);
+		if (memoryContent && memoryOffset <= 32)
+		{
+			bytes contentAsBytes = toBigEndian(*memoryContent);
+			bytes slicedBytes = contentAsBytes | ranges::views::take(unsigned(*memoryOffset)) | ranges::to<vector>;
```

Comment:
```suggestion
			bytes slicedBytes = toBigEndian(*memoryContent) | ranges::views::take(unsigned(*memoryOffset)) | ranges::to<vector>;
```

doesn't get type checked. Probably because it's a range of an rvalue. I wonder if there is a cleaner way to write this.

Reply:
Hm... ``ranges::views::counted(begin(toBigEndian(*memoryContent)), *memoryOffset) | ranges::to<bytes>`` *should* work, but is also not nice...

Reply:
Shouldn't this be `32 - *memoryOffset` (because it is actually "length')?